### DISCUSSION TITLE
New user, but experienced software engineer feedback, and reason not to proceed

### DISCUSSION THREAD

**Discussion by adrian-stephens** [Budibase/budibase#2137] [General]
Hello.   I've been attempting to find an open source MySQL app builder to meet certain requirements. 
To give a bit of background,  a local farm has a huge spreadsheet they use to manage planting and successions.  I offered free of charge to provide them something a bit more stable, based on a database.  I'm not getting paid,  and they have little money,  so something for free is attractive.  I'm a retired software engineer,  with 20+ years of commercial programming real-time systems in C.

I evaluated Zoho creator,  and had a working prototype in a week.   There are other commercial tools out there that would probably do the job.   I also had a partial prototype working in Microsoft Access using ODBC to a MySQL database in a couple of days.

I've looked at Budibase.  My "look" was about 8 hours.   My observations are that the design is nice and clean, and I would really love to use this kind of tool. The tool reflects relationships well,  but only using the inbuilt database.  Importing tables from MySql that include foreign keys (lookup fields) looses that link, and makes the fields nigh on useless.   As the killer reason for a relational database is relations,  that kind of negates the value of an external table feature in budibase.  I wanted to use a MySQL database, because I know how to manage it and could front-end it with MS Access if need be.

Programmability is promised, but unclear.   I need to have a succession form with the following characteristics:
1. select a date range,   fills a subform of beds available based on beds not being allocated to another succession 
during that date range.
2. check/select from that subform of beds available moves them to a subform of beds allocated,  where additional
attributes,  such as the status of the planting/harvesting can be added.
I don't see how to create a trigger based on updating a field in a subform.  I don't see any documentation that explains this.
The documentation page for actions is blank.  There are some promising choices,  but it is not clear how to use them.
I couldn't find how to generate debug messages.

Result,  I don't see how to proceed using this tool without spending a lot of time in trial and error,   which I'm not going to do.

For what it's worth,  I did the same evaluation with the following open-source tools: nuBuilder, Appsmith,  AppGini, Lianja with the same result.

**Comment by arladmin** [comment]
Directus.io might suit your use-case better.

**Comment by mjashanks** [comment]
Hi @adrian-stephens.

Thanks for taking the time to write up such detailed feedback - that is really helpful for us.

Firstly, using SQL relationships. This is about to be released, in the next few days!

Your other questions
1. Are you attempting to filter by date range? I'm not sure what you mean by "fills a subform". I actually just tried to create a demo for you, to filter by a date range, and failed. I logged it here: https://github.com/Budibase/budibase/issues/2138. 
2. In your list of available beds, it should be possible to add a button that updates that particular bed to be allocated on a particular date (or inserts a linked "allocation" row). Happy to tell you more if you're sticking around!

We definitely need better documentation for the reasons you state. This is something that we plan to do a lot more to over the next few weeks and months.

Happy to help you in more detail if you can explain more about what you need here - but it sounds like you're moving on :( 

Thanks again for contributing feedback

**Comment by andy47** [comment]
Has the relationships with external databases been released? I'm not sure which version of Budibase I'm running but I can't seem to see it in my data source from an existing PostgreSQL database.

**Reply by andy47** [reply]
Ok, I did a `docker-compose pull` to update my images and I can now see the 'Relationships' option in my data definition. But I can't tell Budibase about the existing relationships defined via foreign keys, it only lets me create new ones. I don't want (or need) to add new columns to my database as they already exist. All I want to be able to do is to tell Budibase about my existing relationships so that I can use them in my application screens.

**Reply by mjashanks** [reply]
When you create a relationship in budibase, all you are doing is telling budibase which **existing** columns to use for the relationship.

Budibase will not create or modify columns or foreign keys.

**Reply by andy47** [reply]
Ok, thanks for the tips. I think I'm missing where to tell budibase about my existing relationships though. All I have in my data source (PostgreSQL) under Relationships is a button saying 'Create relationship'. When I press this button in my data source I get this dialog box.

![Screen Shot 2021-08-22 at 4 33 21 pm](https://user-images.githubusercontent.com/782758/130345047-364f91e9-28df-43c1-a803-4a1b7f07c0cb.png)

But when I specify the tables to link I get an error message saying 'column name cannot be an existing column'. I'm guessing there is somewhere else for me to define existing relationships but I can't figure out where to do that. Any help is much appreciated.

**Reply by mjashanks** [reply]
Set "Foreign Key" to `currency_id`.

The other 2 options at the bottom are just how the relationship will be named in Budibase, when viewing each of the tables.

**Reply by mjashanks** [reply]
And I think your "from" and  "to" tables should be the other way around

**Reply by andy47** [reply]
Thanks for the quick response, much appreciated. I still don't think I've quite got it right though.

![Screen Shot 2021-08-22 at 4 45 31 pm](https://user-images.githubusercontent.com/782758/130345392-493ab88f-6cd4-4429-b57f-c8e4537fc8fa.png)

**Reply by mjashanks** [reply]
For those two invalid fields, type a name, any name.

It wont create a column in your Db. 🙂

**Reply by andy47** [reply]
Thanks, that did the trick.

**Comment by tvogt** [comment]
Just to second the OP here. I'm coming from PostgreSQL, have done development on Symfony and FileMaker and looking for an alternative to the later.

I found that basics with fields and simple relations work really well, but I'm banging my head against the wall for anything with slightly complex relations or selection. Things that would be a breeze in FileMaker and that I'm sure are possible in Budibase, just I can't find how to do it in the documentation or with a search. Like showing a table with related entries to a currently selected one (what's called Portals in FM). Or filtering a table with a dropdown box. It's all there, I see bits and piece in the docs, but nothing really works for me and it seems a lot more complicated than it should be.
This is also true for external DBs. I'm trying to access an existing table structure and despite clean foreign key relations defined in my Postgres DB, Budibase doesn't pick them up and wants to define its own (redundancy). I'm still trying to wrap my head around things that should be a lot simpler.

I see the potential. But some things are just too arcane or not explained well. I still don't understand what exactly a repeater is and how it's different from a data source or table. Or what needs to be nested into what and why. Is there a difference between nesting a form in a data source vs. the data source in a form?

**Reply by melohagan** [reply]
Thanks for the detailed feedback Tom. I do agree with your points, and I can see what I can do on the documentation side of things to help clarify some of the more complex Budibase specific topics.
 
Kevin is currently working on some Budibase University tutorial videos that does cover Postgres relationships as well as repeaters in more depth which I'm excited for! They should be released fairly soon, and I think users will find them very helpful. 

There may be some options to make the external DB relationships experience smoother. I'll share your feedback with the rest of the dev team to see if they have some ideas!
